### ISSUE TITLE
Saving relationships 

### ISSUE THREAD

**Issue by rwwagner90** [kaliber5/ember-drag-sort#33]
I was wondering if there is an easier way to save the updated lists, other than invoking save on everything? I have this right now:

```js
  @action
  updateColorOrder({ sourceList, sourceIndex, targetList, targetIndex }) {
    if (sourceList === targetList && sourceIndex === targetIndex) return;

    const item = sourceList.objectAt(sourceIndex);

    sourceList.removeAt(sourceIndex);
    targetList.insertAt(targetIndex, item);
    sourceList.invoke('save');
    targetList.invoke('save');
  }
```

It would be ideal to call `save` only on the parents of `sourceList` and `targetList`, if possible. I was thinking perhaps we could support using a target and source model and specifying keys for each list?

**Comment by lolmaus** [comment]
Hmm, my goal was to leave this matter outside of the scope of the addon.

You should be able to use whatever data model you need, and create actual lists as computed properties for `ember-drag-sort` to render them.

If this suggestion does not work for you, please describe the issue in more detail.

**Comment by lolmaus** [comment]
Something like this:

```
class MyComponent extends GlimmerComponent {
  @readOnly('args.sourceData')
  sourceData; // Whatever model you use to represent the lists

  @computed('sourceData.foo.bar.[]')
  get sourceList() {
    // Shape a source list out of the source data
  }

  @computed('sourceData.baz.quux.[]')
  get targetList() {
    // Shape a source list out of the source data
  }

  @action
  updateColorOrder({ sourceList, sourceIndex, targetList, targetIndex }) {
    if (sourceList === targetList && sourceIndex === targetIndex) return;

    this.sourceData.reorder({ sourceList, sourceIndex, targetList, targetIndex });
    this.sourceData.save();
  }
}
```

Note that you can also set the lists to disabled until saving completes/fails.

**Comment by rwwagner90** [comment]
@lolmaus only this addon knows what the lists are. Imagine I have 10 lists, and I drag and drop between 3 and 4. How would my app know what the parents' of those lists are? If instead of passing the arrays as the arguments, we had the option to pass the parent and give the key of the list, we could save the relationships.

**Comment by lolmaus** [comment]
So have something like this?

```
{{#each parents as |parent|}}
  {{#drag-sort-list
    items         = parent.children
    dragEndAction = (action 'dragEnd')
    as |child|
  }}
    {{child.name}}
  {{/drag-sort-list}}
{{/each}}
```

Can't you access the parent from the child?

```
  @action
  updateColorOrder({ sourceList, sourceIndex, targetList, targetIndex }) {
    if (sourceList === targetList && sourceIndex === targetIndex) return;

    const item = sourceList.objectAt(sourceIndex);
    const oldParent = item.parent;

    sourceList.removeAt(sourceIndex);
    targetList.insertAt(targetIndex, item);

    const newParent = item.parent;

    oldParent.save();
    newParent.save();
  }
```

***

If that does not work for you (please explain why), then no, the addon does not currently provide an API to access the parent.

Which approach would be convenient for you? Can you think about some API change that would fix your problem?

**Comment by rwwagner90** [comment]
@lolmaus no, that does not work. It is a many to many relationship. So there is no way to get the parent, since there can be many parents for the child. My suggested change is to allow either arrays or a model, and a key for the array. So then you could do like:

```js
  @action
  updateColorOrder({ sourceModel, sourceKey, sourceIndex, targetModel, targetKey, targetIndex }) {
    if (sourceModel === targetModel && sourceIndex === targetIndex) return;

    const sourceList = sourceModel.get(sourceKey);
    const targetList = targetModel.get(targetKey);
    const item = sourceList.objectAt(sourceIndex);

    sourceList.removeAt(sourceIndex);
    targetList.insertAt(targetIndex, item);

    sourceModel.save();
    targetModel.save();
  }
```

**Comment by lolmaus** [comment]
OK, I see your point, and I admit the case is legit.

How would your suggestion work in the template?

Argument names `sourceModel`, `sourceKey`, etc are very narrow. I'm thinking about using something more general, such as `sourceListArgs` and `targetListArgs`, so that other similar cases can be covered.

**Comment by rwwagner90** [comment]
@lolmaus I am totally open to whatever API you think is best, I just wanted to illustrate my use case.

**Comment by lolmaus** [comment]
Just thinking out loud here...

```
{{#each parents as |parent|}}
  {{#drag-sort-list
    items         = parent.children
    dragEndAction = (action 'dragEnd')
    additionalArgs = (hash
      parent = parent
      key = "children"
    )
    as |child|
  }}
    {{child.name}}
  {{/drag-sort-list}}
{{/each}}
```

And the action could be: 

```
  @action
  updateColorOrder({ sourceList, sourceIndex, sourceArgs, targetList, targetIndex, targetArgs }) {
    if (sourceModel === targetModel && sourceIndex === targetIndex) return;

    const item = sourceList.objectAt(sourceIndex);

    sourceList.removeAt(sourceIndex);
    targetList.insertAt(targetIndex, item);

    sourceArgs.parent.save();
    targetArgs.parent.save();

    console.log(sourceArgs.key); // => "children"
    console.log(targetArgs.key); // => "children"
  }
```

@rwwagner90 What do you think?

**Comment by rwwagner90** [comment]
@lolmaus that looks great to me! 🔥

**Comment by lolmaus** [comment]
@rwwagner90 Wanna give it a shot?

**Comment by rwwagner90** [comment]
@lolmaus I just pulled down the code. Would you be open to ordering things based on the typical ordering from eslint-plugin-ember? https://github.com/ember-cli/eslint-plugin-ember/blob/master/docs/rules/order-in-components.md

It would be easier to contribute if things were a little less custom formatted, in my opinion. I will give it a try though.

**Comment by lolmaus** [comment]
@rwwagner90 Generally yes, but sometimes it is more convenient to group properties by function rather than type.

E. g. having all single line CPs apart from multi line CPs would scatter [isExpanded and isExpanded2](https://github.com/kaliber5/ember-drag-sort/blob/gen-1/addon/components/drag-sort-list.js#L85-L93) (`isExpanded2` is modified in `next` runloop after `isExpanded`). It is much more convenient when these two properties appear together.

For this reason, I'm not enabling the ESLint rule. I find it too rigid and blind to developer needs. Do you think I should reconsider?

IMO, this addon has good code grouping. Is it causing trouble to you?

**Comment by rwwagner90** [comment]
@lolmaus I just find it quite difficult to contribute with things being in a different order with a lot of manual groupings, forcing all the `=` to be evenly spaced, etc. It's 100% your decision, but I tend to like to follow the code style of the default Ember blueprints for ease of contribution.

**Comment by lolmaus** [comment]
@rwwagner90 I've added proper documentation in https://github.com/kaliber5/ember-drag-sort/commit/02e8195d75e94b718d61042505e5827677fb37a4, please review.

**Comment by lolmaus** [comment]
> I just find it quite difficult to contribute with things being in a different order with a lot of manual grouping

This addon started as a personal experiment. I was looking for a simple, Ember-native way to replace `jquery-sortable`. I also used it to fiddle with linting, looking for the most efficient formatting according to my personal taste.

It survived through a couple of unsuccessful attempts before a usable version emerged, then it was adopted by Deveo who let me work on it on paid hours. Then it switched ownership again and is now maintained by kaliber5. But until very recently, I was the only person contributing.

My personal goal regarding code style was to maximize code readability. [StandardJS](https://standardjs.com/) let me remove a lot of visual noise. I started grouping properties in classes way before `order-in-components` became available.

***

I believe there are two big goals:

* Make code style uniform across projects, so that it's easy for developers to switch between them. Automate as much as possible so that developers could focus on results and not the process of coding.
* Let developers stay in charge of formatting code, so that they can manually adjust formatting to improve readability, using their best judgement.

Unfortunately, these two goals often get into contradiction, especially with tools like `order-in-components` and Prettier.

These tools deprive developers from the ability of tweaking code for readability, reducing them to code-writing machines that don't get to write code that's convenient for them.

***

I do *not* mind adopting linting rules that are standard in the community, but:

* I do not want closely related CPs to get scattered solely because one of them is single-line and another one is multi-line. This is ridiculous.

* I don't want the code which I carefully format like this for readability:

    ```js
    const foo =
      bar
        .baz(hmm => hmm.toUpperCase())
        .quux()
        .zomg(zomg => zomg + bar.lol);
    ```

    ...to be force-rewritten like this:

    ```js
    const foo = bar.baz(hmm => hmm.toUpperCase()).quux().zomg(zomg => zomg + bar.lol);
    ```

    Go waste a minute parsing this with your eyes. Do your hourly sacrifice to the lint god.
 
* I don't want to pollute my code with numerous `eslint-disable` and `prettier-ignore` comments.

***

Making huge changes to code style is definitely out of scope of this PR.

And I think we should not make such decisions without getting the current addon's sponsor involved.

But one thing @simonihmig, the CTO of kaliber5, hates is wasting time on bikeshedding. So he would rather enforce Prettier and the default Ember linting ruleset.

Which would probably be a win for you, because you're used to it and to ignoring its disadvantages, but a huge loss for me: as a sentient being and a highly paid specialist, I do not feel comfortable surrendering to dumb automats like Prettier and `order-in-components` that keep uglifying code that I carefully craft without ever considering why it was initially formatted like that in the first place.

***

> forcing all the = to be evenly spaced

This could be auto-formatted by this linting rule: https://github.com/lucasefe/eslint-plugin-align-assignments
